internal/pkg/api/handleCheckin.go

```diff
@@ -274,6 +282,21 @@ func (ct *CheckinT) ProcessRequest(zlog zerolog.Logger, w http.ResponseWriter, r
 	return ct.writeResponse(zlog, w, r, agent, resp)
 }
 
+func (ct *CheckinT) processUpgradeDetails(ctx context.Context, agent *model.Agent, details *UpgradeDetails) error {
```

Comment:
How much additional validation do we need to do here? Do we want to verify that the actionID is valid for the agent? Should we verify that the `metadata` attribute is valid for the `state`?

Reply:
@ycombinator, should we to additional validation for the `upgrade_details` attribute?

Reply:
In general, I would perform the same validation as we do today for the ack API path. So I would imagine that means verifying that the actionID is valid for the agent.

As for whether the `metadata` attribute is valid for the `state`, I think you are referring to this part of the `upgrade_details` object:

```
    "state": "UPG_*",
    "metadata": {
      "scheduled_at": "2023-08-09T10:11:12Z", // when state == "UPG_SCHEDULED"
      "download_percent": 16.4, // when state == "UPG_DOWNLOADING"
      "failed_state": "UPG_*" // when state == "UPG_FAILED"
      "error_msg": "" // when state == "UPG_FAILED"
    }
```

It would be nice to have validation at this finer level but I'd start with just validating `metadata.scheduled_at` and `metadata.error_msg`.  I think for the remaining two metadata properties, we'll need to see whether Agent is capable of setting those reliably.